File: contests/problemC/solutions/solution_fenwick.cpp
#include <bits/stdc++.h>
using namespace std;

struct Fenwick {
  int n;
  vector<int> bit;
  void init(int n_) {
    n = n_;
    bit.assign(n + 1, 0);
  }
  void add(int idx, int val) {
    for (; idx <= n; idx += idx & -idx) bit[idx] += val;
  }
  int prefix(int idx) const {
    if (idx <= 0) return 0;
    if (idx > n) idx = n;
    int res = 0;
    for (; idx > 0; idx -= idx & -idx) res += bit[idx];
    return res;
  }
};

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  long long H, W;
  cin >> H >> W;

  int N;
  if (!(cin >> N)) return 0;

  // tọa độ dùng long long để chịu được tới 1e12
  vector<pair<long long, long long>> cows(N);
  vector<long long> xs;
  xs.reserve(N);

  for (int i = 0; i < N; ++i) {
    cin >> cows[i].first >> cows[i].second; // x, y ≤ 1e12
    xs.push_back(cows[i].first);
  }

  // nén tọa độ x
  sort(xs.begin(), xs.end());
  xs.erase(unique(xs.begin(), xs.end()), xs.end());
  for (auto &c : cows) {
    c.first = lower_bound(xs.begin(), xs.end(), c.first) - xs.begin() + 1; // trong [1..M]
  }

  // sort theo y rồi x; dùng đúng kiểu long long
  sort(cows.begin(), cows.end(),
       [](const pair<long long, long long> &a,
          const pair<long long, long long> &b) {
         if (a.second != b.second) return a.second < b.second;
         return a.first < b.first;
       });

  int M = (int)xs.size();
  Fenwick north, south;
  north.init(M);
  south.init(M);

  // Fenwick chỉ đếm số lượng -> ép về int index
  for (auto &c : cows) north.add((int)c.first, 1);

  int northTotal = N, southTotal = 0;

  auto calc = [&](int pos) {
    int westNorth = north.prefix(pos);
    int westSouth = south.prefix(pos);
    int eastNorth = northTotal - westNorth;
    int eastSouth = southTotal - westSouth;
    return max(max(westNorth, westSouth), max(eastNorth, eastSouth));
  };

  int maxLog = 0;
  while ((1 << (maxLog + 1)) <= M) ++maxLog;

  auto query = [&]() {
    int pos = 0;
    int upperLeft = 0, lowerLeft = 0;
    for (int bitIdx = maxLog; bitIdx >= 0; --bitIdx) {
      int step = 1 << bitIdx;
      if (pos + step > M) continue;
      int npos = pos + step;
      int nextUpper = upperLeft + north.bit[npos];
      int nextLower = lowerLeft + south.bit[npos];
      if (max(nextUpper, nextLower) <=
          max(southTotal - nextLower, northTotal - nextUpper)) {
        upperLeft = nextUpper;
        lowerLeft = nextLower;
        pos = npos;
      }
    }
    int best = calc(pos);
    if (pos + 1 <= M) best = min(best, calc(pos + 1));
    return best;
  };

  int ans = N;
  for (int i = 0; i < N;) {
    int j = i;
    while (j < N && cows[j].second == cows[i].second) {
      int xi = (int)cows[j].first;
      north.add(xi, -1);
      south.add(xi, 1);
      --northTotal;
      ++southTotal;
      ++j;
    }
    ans = min(ans, query());
    i = j;
  }

  cout << ans << '\n';
}

File: contests/problemC/solutions/solution_fenwick.py
import sys
from bisect import bisect_left
from typing import List, Tuple


class Fenwick:
    def __init__(self, n: int) -> None:
        self.n = n
        self.bit = [0] * (n + 1)

    def add(self, idx: int, val: int) -> None:
        while idx <= self.n:
            self.bit[idx] += val
            idx += idx & -idx

    def prefix(self, idx: int) -> int:
        if idx <= 0:
            return 0
        if idx > self.n:
            idx = self.n
        res = 0
        while idx > 0:
            res += self.bit[idx]
            idx -= idx & -idx
        return res


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    it = iter(data)
    _H = int(next(it))
    _W = int(next(it))
    N = int(next(it))

    if N == 0:
        print(0)
        return

    cows: List[List[int]] = []
    xs: List[int] = []
    for _ in range(N):
        x = int(next(it))
        y = int(next(it))
        cows.append([x, y])
        xs.append(x)

    xs.sort()
    unique_xs = []
    prev = None
    for val in xs:
        if val != prev:
            unique_xs.append(val)
            prev = val
    for cow in cows:
        cow[0] = bisect_left(unique_xs, cow[0]) + 1

    cows.sort(key=lambda item: (item[1], item[0]))
    xs_count = len(unique_xs)
    north = Fenwick(xs_count)
    south = Fenwick(xs_count)

    for cow in cows:
        north.add(cow[0], 1)

    north_total = N
    south_total = 0

    def calc(pos: int) -> int:
        west_north = north.prefix(pos)
        west_south = south.prefix(pos)
        east_north = north_total - west_north
        east_south = south_total - west_south
        return max(west_north, west_south, east_north, east_south)

    max_log = 0
    while (1 << (max_log + 1)) <= xs_count:
        max_log += 1

    def query() -> int:
        pos = 0
        upper_left = 0
        lower_left = 0
        for bit_idx in range(max_log, -1, -1):
            step = 1 << bit_idx
            if pos + step > xs_count:
                continue
            npos = pos + step
            next_upper = upper_left + north.bit[npos]
            next_lower = lower_left + south.bit[npos]
            if max(next_upper, next_lower) <= max(south_total - next_lower,
                                                  north_total - next_upper):
                upper_left = next_upper
                lower_left = next_lower
                pos = npos
        best = calc(pos)
        if pos + 1 <= xs_count:
            best = min(best, calc(pos + 1))
        return best

    ans = N
    i = 0
    while i < N:
        j = i
        current_y = cows[i][1]
        while j < N and cows[j][1] == current_y:
            xi = cows[j][0]
            north.add(xi, -1)
            south.add(xi, 1)
            north_total -= 1
            south_total += 1
            j += 1
        ans = min(ans, query())
        i = j

    print(ans)


if __name__ == "__main__":
    main()
